chatkit-core/src/test/kotlin/com/pusher/chatkit/rooms/state/JoinedRoomTest.kt

```diff
@@ -10,55 +11,63 @@ import org.spekframework.spek2.style.specification.describe
 class JoinedRoomTest : Spek({
 
     describe("given no rooms") {
-        val givenState = State(joinedRoomsState = JoinedRoomsState(
-            rooms = emptyMap(),
-            unreadCounts = emptyMap()
-        ))
+        val givenState = State.initial().with(
+            ChatState.initial().with(
```

Comment:
I'm not sure about if this change is for the better. The code ends up needing to know about and implicitly testing what it should not know about or test. Indeed what is already tested in `JoinedRoomsReceivedReducerTest`, i.e. initial (unknown) -> some state transition. The test code should know the state it sets up in the test fixture as "given" and simply calling constructor is cleaner and simpler than two calls `.initial().with`.

Reply:
What I like about it is that the tests will be robust to us adding more fields to the different state objects. Currently they have one member each, so this looks like pointless boilerplate, but if we use the constructors, will we not have to refactor all the tests each time we add a new member field to part of the state?

I really like being able to say in the test "give me an empty (initial) state with this alteration I'm interested in".

Reply:
I suppose we could achieve similar by using the constructors with named params and default values, but then surely we should drop the concept of `initial` methods entirely?

Reply:
OK sorry scrap that. What I was thinking as you've said would require default params and there's no point in repeating logic which is effectively already clearly expressed in `initial()` construction methods. Also having `initial()` with well defined set of `with(...)` gives us opportunity to add some prerequisite checks (like for example assuming that the current user already exists when we start tracking messages or caching other users). Also the test code should not require compromising the design of the code under test. For the production code `initial()` is nice for setting up the store. Sorry I didn't realise calling the constructor is just now that simple as there's only `joinedRoomsState`, thanks for enlightening me as always! 😄